### PR TITLE
Akka.Event: add log filtering system to prevent Akka.NET logs from being emitted in first place

### LINKED ISSUE #7097: Enable debug logging for ShardCoordinator/ShardRegion/Shard actors
**Is your feature request related to a problem? Please describe.**
We are seeing potential issues with the shard coordinator when recovering millions of entities, where shard regions do not receive acknowledgement from the coordinator. We'd like to allow debug logs to be switched on for sharding system-level actors without having to enable debug logging for the whole system

**Describe the solution you'd like**
A configuration switch to perhaps change the log level of all logs from the shard coordinator and shard region actors, so we can view more information, even when the rest of the system is set to `Information` or `Warning` level. 

**Describe alternatives you've considered**
We could switch debug logging on for the whole system but this would cause unnecessary stress on our logging infrastructure.

### PR DESCRIPTION
**Pull request by Aaronontheweb**
## Changes

Fixes #7097 and probably lots of other usability issues.

The goal behind this system is to prevent Akka.NET `LogEvent`s from being emitted in the first place in the event that their data is not useful. This should allow developers to do things like `akka.LOGLEVEL=DEBUG` without getting absolutely flooded with noise from busy parts of the system like Akka.Remote, Akka.Streams, and so on.

The general idea is to pass in a new `Setup` class:

```csharp
/// <summary>
/// Used to specify filters that can be used to curtail noise from sources in the Akka.NET log stream.
/// </summary>
public sealed class LogFilterSetup : Setup
{
    public LogFilterBase[] Filters { get; }
    
    public LogFilterEvaluator CreateEvaluator() => new(Filters);

    public LogFilterSetup(LogFilterBase[] filters)
    {
        Filters = filters;
    }
}
```

These `LogFilterBase` types are intended to be simple functions, but while taking into account some easy performance considerations:

```csharp
public abstract class LogFilterBase : INoSerializationVerificationNeeded, IDeadLetterSuppression
{
    /// <summary>
    /// Which part of the log message this filter is evaluating?
    /// </summary>
    /// <remarks>
    /// This actually has a performance implication - if we're filtering on the source, which
    /// is already fully "expanded" into its final string representation, we can try to fail fast
    /// on that without any additional allocations.
    ///
    /// If we're filtering on the message, we have to fully expand the log message first which
    /// involves allocations. Users on really tight performance budgets should be aware of this.
    /// </remarks>
    public abstract LogFilterType FilterType { get; }

    /// <summary>
    /// Returns <c>true</c> if the message should be kept, <c>false</c> if it should be dropped.
    /// </summary>
    /// <param name="evt">The <see cref="LogEvent"/> being evaluated.</param>
    /// <remarks>
    /// If there are multiple filters, all of them must return <c>true</c> for the message to be kept.
    /// </remarks>
    public abstract bool ShouldKeepMessage(LogEvent evt);

    /// <summary>
    /// Fast path designed to avoid allocating strings if we're filtering on the message content.
    /// </summary>
    /// <param name="part">The part of the message to evaluate.</param>
    /// <param name="content">Usually the fully expanded message content.</param>
    public abstract LogFilterDecision ShouldKeepMessage(LogFilterType part, string content);
}
```

I'm even including a `LogFilterBuilder` that should make it easy to construct these using a simple builder pattern. For a log to be printed, all filters must evaluate to `true`. Will include some examples of this below.

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Design discussion issue #7097
* [x] Changes in public API reviewed, if any.
* [x] I have added website documentation for this feature.

### CODE REVIEW

File: src/core/Akka.API.Tests/verify/CoreAPISpec.ApproveCore.DotNet.verified.txt
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/verify/CoreAPISpec.ApproveCore.DotNet.verified.txt:1728]: New setting - this is where the log filter actually lives. Any custom logging plugin can access it this way.

File: src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs:28]: End to end specs - I can't easily intercept the default output from the `StandardOutWriter` so I'm swapping in my own `MinimalLogger` implementation and testing the results using that at the end. Not 100% clean, but close enough.

File: src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs:97]: Log events will still hit the `EventStream` - the filter runs at the layer beneath this. **AND THAT IS IMPORTANT**: if we have the log filter layer running at this layer, at the point where `LogEvent`s are EMITTED, this WILL KILL PERFORMANCE. 

Instead, we make the decision about whether or not to keep / include a log at the layer right before the log is rendered. **This is the correct place to do this**.

Comment (Aaronontheweb) [reply src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs:97]: Anyway, that's why we still receive all 3 log events.

File: src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs]: The real logger should only have recorded the 1 log that passes both filters.

File: src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/LogFilterEvaluatorSpecs.cs:107]: The rest of these test cases are really just testing the default `Regex` used by the `LogFilterBuilder`.

File: src/core/Akka/Actor/Settings.cs
Comment (Aaronontheweb) [inline src/core/Akka/Actor/Settings.cs:117]: Load the `LogFilterSetup` safely from the `ActorSystemSetup`. **There will be no HOCON equivalent for loading this**.

File: src/core/Akka/Actor/Settings.cs
Comment (Aaronontheweb) [inline src/core/Akka/Actor/Settings.cs:152]: New property added to the `MinimaLogger` - it's settable `internal`ly so we do that here. This is what ties the `LogFilterEvaluator` into the `StandardOutLogger` in default Akka.NET logging scenarios.

File: src/core/Akka/Configuration/akka.conf
Comment (Aaronontheweb) [inline src/core/Akka/Configuration/akka.conf:53]: Disabled the HOCON loading here, as we always `new` up this class by default if this property is left blank. Removes 1 AOT warning that every Akka.NET application 100% hits and improves type safety.

Comment (Aaronontheweb) [reply src/core/Akka/Configuration/akka.conf:53]: You can still override this property if you want to  - I do this in my unit test, which passes.

File: src/core/Akka/Event/LogFilter.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogFilter.cs:19]: **Do not do caching here** - no real utility since we're out of band anyway, and the variety of logs is utterly massive. You want better performance? DON'T EMIT THE LOGS IN THE FIRST PLACE.

Comment (Aaronontheweb) [reply src/core/Akka/Event/LogFilter.cs:19]: LogLevels and granular log emission settings are the right tools to solve performance problems that stem from logging, not the `LogFilter` - the `LogFilter` is here to let people run at more verbose log without keeping all of it.

File: src/core/Akka.Tests/Loggers/LoggerSpec.cs
Comment (Aaronontheweb) [inline src/core/Akka.Tests/Loggers/LoggerSpec.cs:130]: Used the default `NoFilters` placeholder to keep this safe.

File: src/core/Akka/Event/LogFilter.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogFilter.cs:113]: I might get rid of this - because the name of the `ActorSystem` _usually_ gets appended to the `LogSource`, therefore I think this will never work properly for most users. But, you never know.

File: src/core/Akka/Event/LogFilter.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogFilter.cs:159]: Decision-making class on whether or not to include or exclude logs.

File: src/core/Akka/Event/LogFilter.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogFilter.cs:177]: A bit of a perf optimization here - if we're asked to evaluate including / excluding a log based on the content of its message, we have to expand the log earlier than usual. That can actually be a somewhat expensive operation (in the context of millions / second).

Since we have to expand _anyway_ for filtration purposes, we might as well also make this actor responsible for the original log expansion and emit that back out to the `StandardOutLogger` so we only have to do it once. That's why we have the `out` parameter here.

File: src/core/Akka/Event/LogFilter.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogFilter.cs]: Fast path - WE ARE ONLY LOOKING AT LOGSOURCES, which don't need to be expanded. This is great. This is good. We are hoping we hit a negative case and can bail out right away. No expansion of the log is needed at all when this happens.

File: src/core/Akka/Event/LogFilter.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogFilter.cs:194]: Scenario where we have to consider at least 1 rule pertaining to the body of the log message - unavoidable allocation here even in scenarios where we drop the log.

File: src/core/Akka/Event/LogFilter.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogFilter.cs]: A single vote cast for `LogFilterDecision.Drop` is all it takes for us to bail out.

File: src/core/Akka/Event/LogFilter.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/LogFilter.cs:211]: Have to expand the log down here, since we're keeping it.

File: src/core/Akka/Event/StandardOutLogger.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/StandardOutLogger.cs:17]: Default setting, which maintains previous behavior.

File: src/core/Akka/Event/StandardOutLogger.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/StandardOutLogger.cs:128]: Self-explanatory.

File: src/core/Akka/Event/StandardOutLogger.cs
Comment (Aaronontheweb) [inline src/core/Akka/Event/StandardOutLogger.cs:152]: Avoid expanding the log a second time.

**Review by Aaronontheweb — COMMENTED** [review]
Detailed my changes

**Comment by Aaronontheweb** [comment]
Going to leave this PR open for comments for a while before it gets merged in

**Comment by Aaronontheweb** [comment]
Also: documentation. Will add it once I get review comments.

**Review by Arkatufus — APPROVED** [review]
LGTM

**Comment by Aaronontheweb** [comment]
I still need to do API approvals here

**Pull request merged**
